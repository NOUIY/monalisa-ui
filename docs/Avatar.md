# Avatar

A component that represents a user.

**Contents:**
- Avatar
- Avatar Group

<p align="center">
  <img
	  src="https://raw.githubusercontent.com/tuantvk/monalisa-ui/master/assets/monalisaui-avatar.png"
		alt="MonalisaUI"
		width="260">
</p>

View full example: [Avatar Example](https://github.com/tuantvk/monalisa-ui/blob/master/example/Avatar/index.js)

## Avatar

```javascript
import React from 'react';
import { Avatar } from 'monalisa-ui';

const Example = () => (
  <Avatar
    source={{ uri: 'https://picsum.photos/id/434/200/200' }}
  />
);

export default Example;
```

#### Configuration

| Property      | Default       | Option    | Description  |
| ------------- |:-------------:|:---------:|:------------:|
| circle        | false         | boolean   | Shapes the avatar to a circle |
| rounded       | false         | boolean   | Renders avatar with slightly round shaped edges |
| style         | -             | -         | Style for the avatar |
| isVisible     | true          | boolean   | Display avatar background gray |
| width         | 50            | number    | Width of the avatar |
| height        | 50            | number    | Height of the avatar |
| value         | -             | string    | Show value in avatar |


#### Example

```javascript
import React from 'react';
import { View  } from 'react-native';
import { Avatar } from 'monalisa-ui';

const Example = () => (
  <View>
    <Avatar
      source={{ uri: 'https://picsum.photos/id/434/200/200' }}
      width={80}
      height={80}
      circle
    />
    <Avatar
      source={{ uri: 'https://picsum.photos/id/434/200/200' }}
      width={80}
      height={80}
      rounded
    />
    <Avatar
      source={{ uri: 'https://picsum.photos/id/434/200/200' }}
      style={{ marginLeft: 20 }}
    />
  </View>
);

export default Example;
```

## Avatar Group

```javascript
import React from 'react';
import { AvatarGroup } from 'monalisa-ui';

const groups = [
  { source: { uri: 'https://picsum.photos/id/434/200/200' } },
  { source: { uri: 'https://picsum.photos/id/434/200/200' } },
  { source: { uri: 'https://picsum.photos/id/434/200/200' } },
  { source: { uri: 'https://picsum.photos/id/434/200/200' } },
  { source: { uri: 'https://picsum.photos/id/434/200/200' } },
];

const Example = () => (
  <AvatarGroup
    data={groups}
  />
);

export default Example;
```

#### Configuration

| Property      | Default       | Option      | Description  |
| ------------- |:-------------:|:-----------:|:------------:|
| data          | -             | array       | Array of avatar for component **(required)** |
| appearance    | stack         | stack, grid | Indicates the shape of the avatar |
| maxCount      | 0             | number      | The maximum number of avatars allowed in the grid |


#### Example

- stack

```javascript
import React from 'react';
import { AvatarGroup } from 'monalisa-ui';

const groups = [
  { source: { uri: 'https://picsum.photos/id/434/200/200' } },
  { source: { uri: 'https://picsum.photos/id/434/200/200' } },
  { source: { uri: 'https://picsum.photos/id/434/200/200' } },
  { source: { uri: 'https://picsum.photos/id/434/200/200' } },
  { source: { uri: 'https://picsum.photos/id/434/200/200' } },
];

const Example = () => (
  <AvatarGroup
    data={groups}
  />
);

export default Example;
```

- grid

```javascript
import React from 'react';
import { AvatarGroup } from 'monalisa-ui';

const groups = [
  { source: { uri: 'https://picsum.photos/id/434/200/200' } },
  { source: { uri: 'https://picsum.photos/id/434/200/200' } },
  { source: { uri: 'https://picsum.photos/id/434/200/200' } },
  { source: { uri: 'https://picsum.photos/id/434/200/200' } },
  { source: { uri: 'https://picsum.photos/id/434/200/200' } },
];

const Example = () => (
  <AvatarGroup
    data={groups}
    appearance="grid"
  />
);

export default Example;
```

- max count

```javascript
import React from 'react';
import { AvatarGroup } from 'monalisa-ui';

const groups = [
  { source: { uri: 'https://picsum.photos/id/434/200/200' } },
  { source: { uri: 'https://picsum.photos/id/434/200/200' } },
  { source: { uri: 'https://picsum.photos/id/434/200/200' } },
  { source: { uri: 'https://picsum.photos/id/434/200/200' } },
  { source: { uri: 'https://picsum.photos/id/434/200/200' } },
];

const Example = () => (
  <AvatarGroup
    data={groups}
    appearance="grid"
    maxCount={4}
  />
);

export default Example;
```