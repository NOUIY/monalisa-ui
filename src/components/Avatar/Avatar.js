import React from 'react';
import {
  Image,
  View,
  StyleSheet,
  Text,
} from 'react-native';
import { gray_light } from '../../styles';

export default ({
  circle,
  rounded,
  style,
  isVisible = true,
  width = 50,
  height = 50,
  ...rest
}) => {

  // format style avatar
  let customs = [
    { width, height },
    circle && { borderRadius: width / 2 },
    rounded && { borderRadius: width / 10 },
    isVisible ? {} : { ...styles.visible, borderRadius: width / 2 },
    style,
  ];

  if (isVisible) {
    return (
      <Image
        {...rest}
        style={customs}
      />
    );
  }
  return (
    <View style={customs}>
      <Text>+12</Text>
    </View>
  );
}

const styles = StyleSheet.create({
  visible: {
    backgroundColor: gray_light,
    alignItems: 'center',
    justifyContent: 'center',
  }
});