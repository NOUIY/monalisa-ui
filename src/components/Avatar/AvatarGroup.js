/**
 * @license
 * Copyright tuantvk. All Rights Reserved.
 * Licensed under the MIT License. See License in the project root for license information.
 */

import React, { Component } from 'react';
import {
  View,
  StyleSheet,
} from 'react-native';
import Avatar from './Avatar';
import {
  scale,
  subNumber,
} from '../../utils';
import PropTypes from 'prop-types';

class AvatarGroup extends Component {

  _avatarItem = (index, appearance, d) => (
    <Avatar
      circle
      key={index}
      {...d}
      style={[
        appearance === 'stack' && { left: scale(index * -12) },
        appearance === 'grid' && {
          marginHorizontal: scale(3),
          marginBottom: scale(5),
        },
      ]}
    />
  );

  render() {
    const {
      data,
      appearance,
      maxCount,
    } = this.props;

    return (
      <View style={styles.stack}>
        {
          data.map((d, index) => {
            if (maxCount && maxCount !== 0) {
              if (index < maxCount) {
                return this._avatarItem(index, appearance, d);
              }
              if (index === maxCount) {
                return (
                  <Avatar
                    key={index}
                    isVisible={false}
                    value={`+${subNumber(data.length, maxCount)}`}
                    style={appearance === 'stack' && { left: scale(index * -12) }}
                  />
                );
              }
              return null;
            }
            return this._avatarItem(index, appearance, d);
          })
        }
      </View>
    );
  }
}

const styles = StyleSheet.create({
  stack: {
    flexDirection: 'row',
    flexWrap: 'wrap',
  },
});

AvatarGroup.propTypes = {
  data: PropTypes.array.isRequired,
  appearance: PropTypes.oneOf(['stack', 'grid']),
  maxCount: PropTypes.number,
}

AvatarGroup.defaultProps = {
  data: [],
  appearance: 'stack',
  maxCount: 0,
}

export default AvatarGroup;