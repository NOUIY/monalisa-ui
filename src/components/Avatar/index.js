import React from 'react';
import {
  Image,
} from 'react-native';

export default ({
  circle,
  rounded,
  style,
  width = 50,
  height = 50,
  ...rest
}) => (
    <Image
      {...rest}
      style={[
        { width, height },
        circle && { borderRadius: width / 2 },
        rounded && { borderRadius: width / 10 },
        style,
      ]}
    />
  );