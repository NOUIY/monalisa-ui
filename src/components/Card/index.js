import Card from './Card';
import CardImage from './CardImage';

export {
  Card,
  CardImage,
}