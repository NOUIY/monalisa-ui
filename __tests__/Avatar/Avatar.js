import React from 'react';
import { Avatar } from '../../src/components/Avatar';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

describe('Avatar Componnent', () => {

  it('should render without issues', () => {
    const component = shallow(
      <Avatar
        source={{ uri: 'https://picsum.photos/id/434/200/200' }}
      />
    );
    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should pass circle props should still work', () => {
    const component = shallow(
      <Avatar
        circle
        source={{ uri: 'https://picsum.photos/id/434/200/200' }}
      />
    );
    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should pass circle props should still work', () => {
    const component = shallow(
      <Avatar
        rounded
        source={{ uri: 'https://picsum.photos/id/434/200/200' }}
      />
    );
    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should pass style props should still work', () => {
    const component = shallow(
      <Avatar
        rounded
        style={{ width: 100, height: 100 }}
        source={{ uri: 'https://picsum.photos/id/434/200/200' }}
      />
    );
    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });


});