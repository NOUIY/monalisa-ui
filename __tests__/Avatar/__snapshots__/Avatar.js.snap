// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar Componnent should pass circle props should still work 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;

exports[`Avatar Componnent should pass circle props should still work 2`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;

exports[`Avatar Componnent should pass style props should still work 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;

exports[`Avatar Componnent should render without issues 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
